File: src/inception-scheduler/src/main/java/digital/inception/scheduler/service/SchedulerService.java
/*
 * Copyright Marcus Portmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package digital.inception.scheduler.service;

import digital.inception.core.service.InvalidArgumentException;
import digital.inception.core.service.ServiceUnavailableException;
import digital.inception.scheduler.model.DuplicateJobException;
import digital.inception.scheduler.model.Job;
import digital.inception.scheduler.model.JobExecutionFailedException;
import digital.inception.scheduler.model.JobNotFoundException;
import digital.inception.scheduler.model.JobStatus;
import java.util.List;
import java.util.Optional;

/**
 * The {@code SchedulerService} interface defines the functionality provided by a Scheduler Service
 * implementation.
 *
 * @author Marcus Portmann
 */
@SuppressWarnings("unused")
public interface SchedulerService {

  /**
   * Create the job.
   *
   * @param job the {@code Job} instance containing the information for the job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws DuplicateJobException if the job already exists
   * @throws ServiceUnavailableException if the job could not be created
   */
  void createJob(Job job)
      throws InvalidArgumentException, DuplicateJobException, ServiceUnavailableException;

  /**
   * Delete the job
   *
   * @param jobId the ID for the job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobNotFoundException if the job could not be found
   * @throws ServiceUnavailableException if the job could not be deleted
   */
  void deleteJob(String jobId)
      throws InvalidArgumentException, JobNotFoundException, ServiceUnavailableException;

  /**
   * Execute the job.
   *
   * @param job the job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobExecutionFailedException if the job execution failed
   * @throws ServiceUnavailableException if the job could not be executed
   */
  void executeJob(Job job)
      throws InvalidArgumentException, JobExecutionFailedException, ServiceUnavailableException;

  /**
   * Retrieve the job.
   *
   * @param jobId the ID for the job
   * @return the job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobNotFoundException if the job could not be found
   * @throws ServiceUnavailableException if the job could not be retrieved
   */
  Job getJob(String jobId)
      throws InvalidArgumentException, JobNotFoundException, ServiceUnavailableException;

  /**
   * Retrieve the name of the job.
   *
   * @param jobId the ID for the job
   * @return the name of the job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobNotFoundException if the job could not be found
   * @throws ServiceUnavailableException if the name of the job could not be retrieved
   */
  String getJobName(String jobId)
      throws InvalidArgumentException, JobNotFoundException, ServiceUnavailableException;

  /**
   * Retrieve the filtered jobs.
   *
   * @param filter the filter to apply to the jobs
   * @return the filtered jobs
   * @throws ServiceUnavailableException if the filtered jobs could not be retrieved
   */
  List<Job> getJobs(String filter) throws ServiceUnavailableException;

  /**
   * Retrieve the jobs.
   *
   * @return the jobs
   * @throws ServiceUnavailableException if the jobs could not be retrieved
   */
  List<Job> getJobs() throws ServiceUnavailableException;

  /**
   * Returns the maximum number of times the execution of a job will be attempted.
   *
   * @return the maximum number of times the execution of a job will be attempted
   */
  int getMaximumJobExecutionAttempts();

  /**
   * Retrieve the next job that is scheduled for execution.
   *
   * <p>The job will be locked to prevent duplicate processing.
   *
   * @return an Optional containing the next job that is scheduled for execution or an empty
   *     Optional if no jobs are currently scheduled for execution
   * @throws ServiceUnavailableException if the next job scheduled for execution could not be
   *     retrieved
   */
  Optional<Job> getNextJobScheduledForExecution() throws ServiceUnavailableException;

  /**
   * Retrieve the unscheduled jobs.
   *
   * @return the unscheduled jobs
   * @throws ServiceUnavailableException if the unscheduled jobs could not be retrieved
   */
  List<Job> getUnscheduledJobs() throws ServiceUnavailableException;

  /**
   * Reschedule the job for execution.
   *
   * @param jobId the ID for the job
   * @param schedulingPattern the cron-style scheduling pattern for the job used to determine the
   *     next execution time
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobNotFoundException if the job could not be found
   * @throws ServiceUnavailableException if the job could not be rescheduled for execution
   */
  void rescheduleJob(String jobId, String schedulingPattern)
      throws InvalidArgumentException, JobNotFoundException, ServiceUnavailableException;

  /**
   * Reset the job locks.
   *
   * @param status the current status of the jobs that have been locked
   * @param newStatus the new status for the jobs that have been unlocked
   * @throws InvalidArgumentException if an argument is invalid
   * @throws ServiceUnavailableException if the job locks could not be reset
   */
  void resetJobLocks(JobStatus status, JobStatus newStatus)
      throws InvalidArgumentException, ServiceUnavailableException;

  /**
   * Schedule the next unscheduled job for execution.
   *
   * @return {@code true} if a job was successfully scheduled for execution or {@code false}
   *     otherwise
   * @throws ServiceUnavailableException if the next unscheduled job could not be scheduled for
   *     execution
   */
  boolean scheduleNextUnscheduledJobForExecution() throws ServiceUnavailableException;

  /**
   * Set the status for the job.
   *
   * @param jobId the ID for the job
   * @param status the new status for the job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobNotFoundException if the job could not be found
   * @throws ServiceUnavailableException if the status could not be set for the job
   */
  void setJobStatus(String jobId, JobStatus status)
      throws InvalidArgumentException, JobNotFoundException, ServiceUnavailableException;

  /**
   * Unlock a locked job.
   *
   * @param jobId the ID for the job
   * @param status the new status for the unlocked job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobNotFoundException if the job could not be found
   * @throws ServiceUnavailableException if the job could not be unlocked
   */
  void unlockJob(String jobId, JobStatus status)
      throws InvalidArgumentException, JobNotFoundException, ServiceUnavailableException;

  /**
   * Update the job.
   *
   * @param job the {@code Job} instance containing the updated information for the job
   * @throws InvalidArgumentException if an argument is invalid
   * @throws JobNotFoundException if the job could not be found
   * @throws ServiceUnavailableException if the job could not be updated
   */
  void updateJob(Job job)
      throws InvalidArgumentException, JobNotFoundException, ServiceUnavailableException;
}

File: src/inception-angular/projects/ngx-inception/scheduler/src/services/scheduler.service.ts
/*
 * Copyright Marcus Portmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  HttpClient,
  HttpErrorResponse,
  HttpResponse
} from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import {
  AccessDeniedError,
  CommunicationError,
  INCEPTION_CONFIG,
  InceptionConfig,
  InvalidArgumentError,
  ProblemDetails,
  ResponseConverter,
  ServiceUnavailableError
} from 'ngx-inception/core';
import { Observable, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { Job } from './job';
import { JobStatus } from './job-status';
import {
  DuplicateJobError,
  JobNotFoundError
} from './scheduler.service.errors';

/**
 * The Scheduler Service implementation.
 *
 * @author Marcus Portmann
 */
@Injectable({
  providedIn: 'root'
})
export class SchedulerService {
  /**
   * Constructs a new SchedulerService.
   *
   * @param config     The Inception configuration.
   * @param httpClient The HTTP client.
   */
  constructor(
    @Inject(INCEPTION_CONFIG) private config: InceptionConfig,
    private httpClient: HttpClient
  ) {
    console.log('Initializing the Scheduler Service');
  }

  /**
   * Retrieve the description for the job status.
   *
   * @param jobStatus the job status
   */
  static getJobStatusDescription(jobStatus: JobStatus): string {
    switch (jobStatus) {
      case JobStatus.Unscheduled:
        return $localize`:@@scheduler_job_status_unscheduled:Unscheduled`;
      case JobStatus.Scheduled:
        return $localize`:@@scheduler_job_status_scheduled:Scheduled`;
      case JobStatus.Executing:
        return $localize`:@@scheduler_job_status_executing:Executing`;
      case JobStatus.Executed:
        return $localize`:@@scheduler_job_status_executed:Executed`;
      case JobStatus.Aborted:
        return $localize`:@@scheduler_job_status_aborted:Aborted`;
      case JobStatus.Failed:
        return $localize`:@@scheduler_job_status_failed:Failed`;
      case JobStatus.OnceOff:
        return $localize`:@@scheduler_job_status_once_off:Once Off`;
      default:
        return $localize`:@@scheduler_job_status_unknown:Unknown`;
    }
  }

  /**
   * Create a job.
   *
   * @param job The job to create.
   *
   * @return True if the job was created successfully or false otherwise.
   */
  createJob(job: Job): Observable<boolean> {
    return this.httpClient
      .post<boolean>(`${this.config.apiUrlPrefix}/scheduler/jobs`, job, {
        observe: 'response'
      })
      .pipe(
        map(SchedulerService.isResponse204),
        catchError(SchedulerService.handleApiError('Failed to create the job.'))
      );
  }

  /**
   * Delete the job.
   *
   * @param jobId The ID for the job.
   *
   * @return True if the job was deleted or false otherwise.
   */
  deleteJob(jobId: string): Observable<boolean> {
    return this.httpClient
      .delete<boolean>(
        `${this.config.apiUrlPrefix}/scheduler/jobs/${encodeURIComponent(jobId)}`,
        { observe: 'response' }
      )
      .pipe(
        map(SchedulerService.isResponse204),
        catchError(SchedulerService.handleApiError('Failed to delete the job.'))
      );
  }

  /**
   * Retrieve the job.
   *
   * @param jobId The ID for the job.
   *
   * @return The job.
   */
  @ResponseConverter getJob(jobId: string): Observable<Job> {
    return this.httpClient
      .get<Job>(
        `${this.config.apiUrlPrefix}/scheduler/jobs/${encodeURIComponent(jobId)}`,
        { reportProgress: true }
      )
      .pipe(
        catchError(
          SchedulerService.handleApiError('Failed to retrieve the job.')
        )
      );
  }

  /**
   * Retrieve the name of the job.
   *
   * @param jobId The ID for the job.
   *
   * @return The name of the job.
   */
  getJobName(jobId: string): Observable<string> {
    return this.httpClient
      .get<string>(
        `${this.config.apiUrlPrefix}/scheduler/jobs/${encodeURIComponent(jobId)}/name`,
        { reportProgress: true }
      )
      .pipe(
        catchError(
          SchedulerService.handleApiError('Failed to retrieve the job name.')
        )
      );
  }

  /**
   * Retrieve all the jobs.
   *
   * @return The jobs.
   */
  @ResponseConverter getJobs(): Observable<Job[]> {
    return this.httpClient
      .get<
        Job[]
      >(`${this.config.apiUrlPrefix}/scheduler/jobs`, { reportProgress: true })
      .pipe(
        catchError(
          SchedulerService.handleApiError('Failed to retrieve the jobs.')
        )
      );
  }

  /**
   * Update the job.
   *
   * @param job The job.
   *
   * @return True if the job was updated successfully or false otherwise.
   */
  updateJob(job: Job): Observable<boolean> {
    return this.httpClient
      .put<boolean>(
        `${this.config.apiUrlPrefix}/scheduler/jobs/${encodeURIComponent(job.id)}`,
        job,
        { observe: 'response' }
      )
      .pipe(
        map(SchedulerService.isResponse204),
        catchError(SchedulerService.handleApiError('Failed to update the job.'))
      );
  }

  private static handleApiError(defaultMessage: string) {
    return (httpErrorResponse: HttpErrorResponse) => {
      if (
        ProblemDetails.isProblemDetails(
          httpErrorResponse,
          DuplicateJobError.TYPE
        )
      ) {
        return throwError(() => new DuplicateJobError(httpErrorResponse));
      } else if (
        ProblemDetails.isProblemDetails(
          httpErrorResponse,
          JobNotFoundError.TYPE
        )
      ) {
        return throwError(() => new JobNotFoundError(httpErrorResponse));
      } else if (AccessDeniedError.isAccessDeniedError(httpErrorResponse)) {
        return throwError(() => new AccessDeniedError(httpErrorResponse));
      } else if (CommunicationError.isCommunicationError(httpErrorResponse)) {
        return throwError(() => new CommunicationError(httpErrorResponse));
      } else if (
        InvalidArgumentError.isInvalidArgumentError(httpErrorResponse)
      ) {
        return throwError(() => new InvalidArgumentError(httpErrorResponse));
      }

      return throwError(
        () => new ServiceUnavailableError(defaultMessage, httpErrorResponse)
      );
    };
  }

  /**
   * Helper method to check if the response status is 204.
   *
   * @param httpResponse The HTTP response.
   * @return True if the status is 204, otherwise false.
   */
  private static isResponse204(httpResponse: HttpResponse<boolean>): boolean {
    return httpResponse.status === 204;
  }
}
